### DISCUSSION TITLE
Supporting streaming SSR in the future

### DISCUSSION THREAD

**Discussion by Rich-Harris** [sveltejs/kit#4831] [Ideas]
Currently, SvelteKit renders HTML in one go, which I tend to think is the better choice despite that being a minority opinion (you can't set the response status code once you've started sending the body, and you generally don't know for sure what the status code is _until_ you've rendered the body, so AFAICT streaming implies 200-everything-and-hope-for-the-best).

It might be worth preparing for a world in which a) we want to give people the option to stream, and b) it's possible to set status codes late (which I understand is under active discussion), in which case streaming HTML makes more sense to me.

There are currently three obstacles that I'm aware of:

## Svelte

Svelte itself doesn't have a streaming render mode, and without it there's not much benefit to streaming the non-Svelte parts of a page (other than optimising TTFB, I guess, which isn't a particularly meaningful achievement). That could change — we could introduce APIs that allowed us to (for example) resolve `{#await}` blocks rather than rendering the pending state:

```js
Component.stream(
  (error, chunk) => {
    if (error) {
      // literally no idea what to do here tbh
    } else {
      controller.enqueue(chunk);
    }
  },
  {
    timeout: 1000
  }
);
```

In SvelteKit, we could wrap nested routes in `{#await}` blocks that wait for `load` functions to resolve, allowing us to render the nav bar before the main content is ready, etc.

> Returning a `ReadableStream` directly from `Component.stream` feels more standards-flavoured, but I'm not sure it would actually be an ergonomic improvement. I reserve the right to change my opinion in future, but for now it's an immaterial detail

## `<svelte:head>`

If we stream everything in order, [`<svelte:head>`](https://svelte.dev/docs#template-syntax-svelte-head) puts us back to square one — we can't render the `%svelte.head%` in `src/app.html` until we've finished rendering the body and know that we won't encounter any more `<svelte:head>` elements.

Two options spring to mind: we somehow gather head content prior to rendering, or we stream out of order. The first option makes sense in some cases, but not all (e.g. a document title often depends on the same data as the body content), and would involve some fairly substantial design changes.

The second option would be easier (render a placeholder element to the head, and write a small `<script>` that inserts HTML adjacent to the placeholder as `<svelte:head>` elements are rendered), though out-of-order streaming makes me a little nervous — it relies on JavaScript (albeit not in the inherently brittle way that 'relies on JavaScript' usually implies), and I would expect it to be significantly worse for SEO, unless we did something tricksy like defaulting to in-order streaming if the UA is a known search engine bot.

## `transformPage`

People use `transformPage` for making arbitrary tweaks to the HTML (like replacing a `<html lang="%lang%">` placeholder), and the API assumes the whole page is rendered.

In a streaming world this API makes no sense. We need to transform the stream instead, which in the general case is [rather more complicated](https://github.com/jakearchibald/isserviceworkerready/blob/master/src/demos/transform-stream/edge-cases.md) than something like a naive `transformPageChunk`.

---

This is all fairly vague, because I don't have a lot of answers or strong opinions. I'm mostly putting this here so we have a place for people who _do_ have answers and strong opinions (including about how important streaming is).

**Comment by Mlocik97** [comment]
I find it complicating things too... it's more prone to bugs and unexpected behavior.

On the other size I find streaming as good feature for files that are typically "big". (Which in case of code spliting, is not that much problem for code). Instead of doing drastic change at once, I would probably add streaming for large files (that are not critical code, maybe for non text format files) first (we can maybe check for size before we serve it?), and then see what result we get. Maybe there could be option for it in `config.kit`. Also we don't really need to stream it by "bytes" but wouldn't it be possible to stream it by "lines" in case of HTML, CSS, JS and other text format files (we would just need to check for `\n` character and line number), but yeah, this give us question about minifing?

**Reply by Rich-Harris** [reply]
This discussion is purely about streaming pages — endpoint request/response streams is covered by https://github.com/sveltejs/kit/issues/3419

**Reply by Mlocik97** [reply]
oh, my bad... for pages I'm not sure if streaming doesn't add more problems than solve...

**Comment by Rich-Harris** [comment]
Some useful tips I was given:

* [`<meta http-equiv="Status" content="500">`](https://indieweb.org/meta_http-equiv_status) — apparently this is a way to set the status of a document within the markup (unclear if it works if it's inside `<body>`, which it would need to be, or whether it actually overrides the response status or is merely a hint for servers dealing with static HTML. also, googlebot ignores it)
* `<meta name="robots" content="noindex" />` (which can be in the body) will prevent error pages from being indexed by search engines, which I guess in practice is the main benefit of attaching status codes to pages. So perhaps that's the way to make peace with 200-everything-and-hope-for-the-best

**Comment by Rich-Harris** [comment]
Another challenge to streaming is that we can't start rendering the root layout component until the leaf component's data has loaded, because the layout may depend on `stuff` that is set by the leaf. We could maybe do some janky static analysis to detect when waiting is unnecessary, but I'm not sure the complexity justifies itself.

**Reply by Mlocik97** [reply]
When I read and think about this problem, I really feel like streaming pages adds more problems than solve. And basically pages are typically not that big (hope no one makes pages with so a lot of HTML nodes as W3C in their specs) that streaming of pages would be "must have". In the end I think it's good to implement streaming for endpoint request/responses but not for pages.

**Reply by Rich-Harris** [reply]
Size isn't the issue, it's whether you can start sending useful bytes before you've fetched all the data your page needs

**Comment by benmccann** [comment]
What might make a lot more sense to implement and perhaps get even more performance gains would be HTTP 103 Early Hints. I think we could send the list of JS / CSS files we expect the page will need early on and then respond later with the SSR-rendered HTML.

**Reply by Rich-Harris** [reply]
I don't think Early Hints is within our control - it doesn't fit the Request/Response model. Best we can do is render Link headers. I think we _should_ be able to do that as soon as route matching is complete, without actually rendering, which would be a substantial win by itself but requires us to change the `transformPage` API

**Reply by Rich-Harris** [reply]
Further context: there's a PR that partially implements Link headers (it waits until rendering though, limiting its benefit) but it was closed prematurely: https://github.com/sveltejs/kit/pull/4785

**Reply by benmccann** [reply]
> I don't think Early Hints is within our control - it doesn't fit the Request/Response model. Best we can do is render Link headers. I think we should be able to do that as soon as route matching is complete, without actually rendering, which would be a substantial win by itself but requires us to change the `transformPage` API

This sounds pretty close to using early hints, but I think it might have to be implemented as hints still rather than just sending the headers early. The reason I say that is for error handling. I think the status code needs to be sent first and you don't know what the status code is until the page is rendered (e.g. what if it errors and you need to send a 500?). With early hints we can stream a 103 response and the `Link` headers, hold the stream open, then when we get the `Response` we can serialize it to the stream and send it. This allows us to send the `Link` headers as soon as the route is matched before doing the actual page rendering, which I think we couldn't otherwise do since we wouldn't know the response code yet.

I think it still fits with the `Request`/`Response` model. We still have that as the API exposed to SvelteKit users. Behind the scenes we'll be opening a stream and sending a 103 before getting the `Response` from the user. And then when we do get the `Response` we can send that on the open stream. I do wonder if there's some performance impact here of holding open more connections, so perhaps we should allow users to toggle it on and off.

**Reply by benmccann** [reply]
Answering my own question... I just realized the `Response` API from `fetch` might not make this easy to do, so this is hard on the serverless platforms.

Our current `Server` API doesn't make this very easy since it relies on `Response` which doesn't support early hints. Perhaps we can change the `Server` API to make this work for `adapter-node` at least. The API might not be as elegant though

**Reply by benmccann** [reply]
It's not entirely clear to me you could do this with Polka in `adapter-node` today as the Node.js APIs don't really support it (https://github.com/nodejs/node/issues/27921). It does look like we could at least make it work if we switched from Polka to Fastify which lets you do it by dropping down to the raw socket (https://github.com/fastify/fastify/issues/2683#issuecomment-726106616).

**Reply by Rich-Harris** [reply]
Exactly — my point is that in the general case, Early Hints is outside our control, since all the platforms are moving towards browser-API-centric `Request` -> `Response` models (whether that's exposed via a `handle` export or a service worker API or whatever). It's possible that the `Response` API could evolve to accommodate early hints...

```js
const response = new Response(undefined, {
  status: 103,
  headers: { link }
});

render_page().then(page => {
  response.setFinalStatus(page.status);
  response.setFinalHeaders(page.headers);
  response.setBody(page.body); // could be a ReadableStream
});

return response;
```

...but that's far in the future, given that the [Early Hints RFC](https://datatracker.ietf.org/doc/html/rfc8297) itself is still 'experimental' (despite being written in 2017 😩).

For now, this would mean 200-everything-and-hope-for-the-best with the tricks described in https://github.com/sveltejs/kit/discussions/4831#discussioncomment-2695379.

I _would_ love to better understand how other frameworks think about this problem, though.

**Reply by Mlocik97** [reply]
5 years and still.... bleh...

and I though `BeforeInstallPromptEvent` was too dead to think about it as deprecated feature (that even mozilla mark as one, and has 2 warning in description (Experimental, Non-Standard).

The issue with `200` is what if "connection" is teared in between...

The only two ideas I got are

To try at the very beginning to load script to client-side that will controll other loading, and when issue at streaming happens it will `ping` some Kit's specific endpoint (with empty bodies) that return status 500 and falls to error page (that would be loaded after)? Somehow...  But then that client-side script that controlls loading from stream would not be streamable itself.

Or another better idea use `<meta http-equiv="refresh">` that would be streamed at beginning to head tag, and after everything is loaded from stream, JS code will remove it from head (would be apllicable only if JS is enabled)... this will cause so after specified time if content is not loaded (aka stream connection was teared), it will not be removed, it will refresh to error page with status 500?

`<meta http-equiv="refresh" content="5; url=/error_500">`

But this adds problem about timing...

**Comment by AlbertMarashi** [comment]
I'm personally against the streaming API it doesn't make sense and it doesn't really make sense to use it.

**Reply by andy0130tw** [reply]
For an argument on how streaming API can be useful, see https://dev.to/tigt/the-weirdly-obscure-art-of-streamed-html-4gc2.

**Comment by tigt** [comment]
I spent a lot of time trying to figure out how to best interoperate on this problem, so forgive me for this big comment.

### When is HTML streaming most useful?

I worked on a site where the flow to get a response code looked like:

1. Parse the request. If it’s clearly wrong somehow, emit a `4XX`. (“Is this a valid product URL? Is that a valid UPC?”)
2. Make an API call to see if the request can be answered with good data. (“Does this UPC correspond to a product we have?")
3. Make an API call to get the actual data (“Get details for this product UPC")
4. Make a bunch of other API calls for related products, reviews, and other useful but non-critical content

Streaming was most helpful right after №2. Sites where steps 2 and 3 happen at the same time (like a SQL query) would not benefit as much, but they would still benefit for data fetching like №4.

Another case was search results. A search that returns 0 results is still a `200 OK`, so it was fine to [start streaming search results after step №1](https://dev.to/tigt/skeleton-screens-but-fast-48f1) **if and only if there was a way to indicate a mid-stream error damaged the response**.

### What status information would be helpful mid-stream?

- Anything `1XX` is already mid-response information, so it’s all good
- `2XX` is the success case so we probably don’t have to worry
- `3XX` _can_ contain bodies but nowadays almost never. If you do need a mid-HTML redirect, there’s `<meta http-equiv=refresh content="0;…">` or `<script>location.href = '…'</script>` or both.
- **`4XX` is the missing, important part of this story.** Some aren’t a problem (`400 Bad Request` happens before response processing even begins), but important errors like `404` are.
- `5XX` semantics can be conveyed with stream error mechanisms that vary with the HTTP version

### `5XX` server error semantics

This is the only one developers can consistently signal today. Luckily, “something broke” is the most common reason to signal HTTP-level information mid-HTML:

- The response is non-authoritative and shouldn’t be cached or indexed if it can be helped
- A retry may fix the problem (this is why prior art like Rails’ mid-stream redirects to an error page aren’t user-friendly, since it breaks the Refresh button)
- Whether it’s safe or unsafe to retry

No machine-readable indicator that a streamed response contains erroneous content causes some thorny problems:

- HTTP caches may store the erroneous content and reuse it
- Search engines and other bots will store/index the erroneous content, since they received no sign they should retry or discard the response
- HTTP-level tools (debuggers, monitoring, `curl`, spiders, etc.) will report the response as successful, even though it wasn’t

HTTP/1.1’s chunked responses and HTTP/2’s stream error handling each have a standardized way of indicating a dynamic response failed to complete successfully:

<dl>
  <dt>HTTP/1.1 <code>Transfer-Encoding: chunked</code></dt>
  <dd><a href="https://httpwg.org/http-core/draft-ietf-httpbis-messaging-latest.html#incomplete.messages">IETF Draft: HTTP/1.1 Messaging §8 Handling Incomplete Messages</a></dd>
  <dd><p>If a chunked response doesn’t terminate with the zero-length end chunk, the client must assume that the response was <i>incomplete</i> — which at the very least, <a href="https://httpwg.org/http-core/draft-ietf-httpbis-cache-latest.html#rfc.section.3.1">means a cache should double-check with the server before reusing the stored incomplete response</a>.</p>
  <p>This can be conveyed by either writing incorrect chunk delimiters, or hanging up the TCP stream before the final zero-length chunk.</p>
</dd>

  <dt>HTTP/2 Streams</dt>
  <dd><a href="https://tools.ietf.org/html/rfc7540#section-5.4.2">RFC 7540 §5.4.2 Stream Error Handling</a></dd>
  <dd><p>An HTTP/2 stream can signal an application error by sending a <code>RST_STREAM</code> frame with an error code of <code>0x2 INTERNAL_ERROR</code>.</p></dd>
</dl>

Propagating mid-stream errors to clients may take some doing; at least NGiNX by default assumes incorrectly-terminated Chunked Transfer-Encoding means your backend is a bozo and it should fix it for you. Research needed into how common reverse proxies or load balancers handle error-signaled HTTP streams:

- [Real-world examples of how that can happen, including the comment at the very bottom](https://rhodesmill.org/brandon/2013/chunked-wsgi/)
- [Atmosphere long-polling on nginx – chunked encoding error](https://rijulaggarwal.wordpress.com/2018/01/10/atmosphere-long-polling-on-nginx-chunked-encoding-error/)
- In the common case of a backend speaking HTTP/1.1 to a reverse proxy/front-end terminator/CDN/etc. that translates to HTTP/2 for browsers, what can be done?

#### Other fixes

Historically, I’ve seen other workarounds for this sort of error, but with drawbacks:

<dl>
<dt>For known bots, don't stream at all.
<dd>Unfortunately, you can’t always know it’s a bot. For example, Google can pretend to be a real browser to check you’re not selectively serving them different content.
<dd>Makes TTFB much worse for bots that rank your speed.

<dt>For mid-stream errors, redirect to an error page via `meta[http-equiv=refresh]` or something
<dd>Unsure how well this works across the HTTP ecosystem; tools like `curl` don’t benefit.
<dd>Without an explicit damage signal, the erroneous response may be cached.
<dd>Invites dangling markup problems, depending on where in the HTML the error occurred.
<dd>Also breaks the Refresh button, worsening the user experience.
</dl>

### `4XX` client errors and when not to start streaming

I highly recommended **not to begin streaming** HTML for crucial page content, such as:

- If a product URL addresses a product you have (`200 OK`) or something completely bogus (`404 Not Found`)
- If a user-specific area has valid authentication (`200 OK`) or needs to log in first (`403 Forbidden`)
- If a time-limited promotion has ended (`401 Gone`) or is still ongoing (`200 OK`)

In these cases, I waited to begin the HTTP stream until I know which status code to send.

If tempted to begin streaming ASAP anyway for speed reasons, consider the following:

- The [TCP connection still warms up to grow its window size](https://hpbn.co/building-blocks-of-tcp/#slow-start) while your server thinks about the status code.
- If you want to preload likely assets during server think time, use [the `103 Early Hints` status code](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/103).
- HTML caching can remember the status code and accelerate future queries with the correct status, after a more-expensive first one.

For trivial errors, I did nothing different. If the “recommended for you” section failed to render, I swallowed the error and sent the rest of the page out.

### Further info

- [The Mysteries of Connection Close · HTTP: The Definitive Guide](https://www.oreilly.com/library/view/http-the-definitive/1565925092/ch04s07.html)
- [Apache Bug 55475: mod_proxy breaks HTTP chunked transfer coding](https://bz.apache.org/bugzilla/show_bug.cgi?id=55475)
- I proposed this problem to the HTTPWG, and there’s been efforts, but they need sponsorship to move them through the standards process
  - https://github.com/httpwg/http-core/issues/895#issuecomment-875271020
  - https://github.com/mnot/I-D/blob/400a0c6fc26e059863790cad13e31c808695f39b/Abandoned/draft-nottingham-cache-trailers.md

**Reply by IshanKBG** [reply]
Any news about this?

**Comment by jdgamble555** [comment]
Streaming and Suspense are really the same feature. Linking to the [Suspense](https://github.com/sveltejs/svelte/issues/3203) thread.  From my understanding, all streaming really does is this:

NextJS (for example) seems to just load the component as usual, loading only the `fallback` at first.

```html
<!-- 1. loads the html and head from serverSideProps -->
<html>
<head>
...
</head>
<body>
<!-- 2. loads the body and any suspense fallback components, starts the async fetches in the children... -->
<Suspense fallback="loading...">
  <OtherComponent />
</Suspense>
...
<!-- 3. Selectively hydrates the page, as if complete to allow interaction -->
<!-- 4. loads the script tag at the end with the data to hydrate the component(s) in suspense after fetch completes -->
<script>
// ... the js code to hydrate the "loading" with the correct info in "other component" will be here 
</script>
</body>
<!-- 4. Runs the JS to Selectively hydrates again replacing fallback with actual data -->
```

It seems to me that all it's doing is loading the hydration information last, while allowing you to interact on your page quicker.  It will always wait for the [head information first](https://beta.nextjs.org/docs/data-fetching/streaming-and-suspense#streaming-and-seo), so you can't really stop that.  It is not good if you need to load SEO information later.  It is only good for quicker user interaction.  Hence, we still need load functions (like nextjs's getServerSideProps) to get our data first.

It is streamed because the page does not complete until the JS info is hydrated at the end (remember before we had `defer` in js).

So, @Rich-Harris correct me if I'm wrong, but I don't think NextJS has any of the problems you mentioned, because it is not built for SEO information in the header.  Technically it could help with SEO due to a higher Lighthouse score and body content, but it would have nothing to do with `<head>` indexing. 

It seems the body gets hydrated, and then hydrated again using **Selective Hydration**.  That way the time to interactive is faster.  It is not fetching anything from the client, so it is way faster.

This is all thanks to the [renderToPipeableStream](https://beta.reactjs.org/reference/react-dom/server/renderToPipeableStream) api that React 18 wrote from Node.js Stream.

Links:
- [React Beta Docs - Suspense](https://beta.reactjs.org/reference/react/Suspense)
- [Streaming Server Rendering with Suspense](https://www.youtube.com/watch?v=pj5N-Khihgc) - this video seems to explain it well from the React 2021 Conference

So, I don't see why this would be too difficult in SvelteKit, or at least not impossible.

Just my thoughts,

J
___

Also worth adding, this is how Qwik loads its data by default when using the [useResource$](https://qwik.builder.io/docs/components/resource/) function. It streams the data (confirmed by Qwik).

**Comment by mpiorowski** [comment]
With the new stream feature, aren't we one step closer to make it work? @Rich-Harris ? 

This feature would be AWESOME with gRPC streaming connection, where the data can be visible to the client the moment it cames to the server :)

**Comment by MrBns** [comment]
After 2 years,
does sveltekit has streaming SSR support ? or what is the status on this discussion.

**Reply by jdgamble555** [reply]
https://svelte.dev/blog/streaming-snapshots-sveltekit
